d2/src/test/java/com/linkedin/d2/balancer/servers/ZookeeperConnectionManagerTest.java

```diff
@@ -471,6 +472,297 @@ public void testMarkUpAndDownMultipleTimesFinalUp()
+    UriProperties properties = store.get(_cluster);
+    assertNotNull(properties);
+    assertEquals(properties.getPartitionDataMap(URI.create(_uri)).get(DefaultPartitionAccessor.DEFAULT_PARTITION_ID).getWeight(), WEIGHT);
+    assertEquals(properties.Uris().size(), 1);
+
+    shutdownManager(manager);
+  }
+
+  @Test
+  public void testNoWarmupWhenExecutorServiceIsNull() throws Exception{
+    ScheduledExecutorService warmupExecutorService = null;
+    boolean isDarkWarmupEnabled = true;
+    String warmupClusterName = "warmup" + _cluster;
+    int warmupDuration = 5; //Run warm-up for 5 seconds
+
+    ZooKeeperAnnouncer announcer = getZooKeeperWarmupAnnouncer(_cluster, _uri, WEIGHT, isDarkWarmupEnabled, warmupClusterName, warmupDuration, warmupExecutorService);
+    ZooKeeperConnectionManager manager = createManagerForWarmupTests(true, warmupDuration, announcer);
+    ZooKeeperEphemeralStore<UriProperties> store = createAndStartUriStore();
+
+    // Assert that no warm-up and only mark up to the regular cluster
+    UriProperties properties = store.get(warmupClusterName);
+    assertNull(properties);
+    properties = store.get(_cluster);
+    assertNotNull(properties);
+    assertEquals(properties.getPartitionDataMap(URI.create(_uri)).get(DefaultPartitionAccessor.DEFAULT_PARTITION_ID).getWeight(), WEIGHT);
+    assertEquals(properties.Uris().size(), 1);
+
+    shutdownManager(manager);
+  }
+
+  @Test (invocationCount = 10, timeOut = 15000)
+  public void testWarmup() throws Exception
+  {
+    ScheduledExecutorService warmupExecutorService = Executors.newSingleThreadScheduledExecutor();
+    boolean isDarkWarmupEnabled = true;
+    String warmupClusterName = "warmup" + _cluster;
+    int warmupDuration = 5; //run warm-up for 5 seconds
+
+    ZooKeeperAnnouncer announcer = getZooKeeperWarmupAnnouncer(_cluster, _uri, WEIGHT, isDarkWarmupEnabled, warmupClusterName, warmupDuration, warmupExecutorService);
+    ZooKeeperConnectionManager manager = createManagerForWarmupTests(false, warmupDuration, announcer);
+    ZooKeeperEphemeralStore<UriProperties> store = createAndStartUriStore();
+    FutureCallback<None> managerStartCallback = new FutureCallback<>();
+    manager.start(managerStartCallback);
+
+    // Wait till warm up completes and announcer successfully marks up to the regular cluster
+    managerStartCallback.get();
+
+    UriProperties properties = store.get(_cluster);
+    assertNotNull(properties);
+    assertEquals(properties.getPartitionDataMap(URI.create(_uri)).get(DefaultPartitionAccessor.DEFAULT_PARTITION_ID).getWeight(), WEIGHT);
+    assertEquals(properties.Uris().size(), 1);
+
+    // If warm up has happened, mark down for the warm up cluster should be successful
+    properties = store.get(warmupClusterName);
+    assertNotNull(properties);
+    assertEquals(properties.Uris().size(), 0);
+
+    shutdownManager(manager);
+  }
+
+  @Test (invocationCount = 10, timeOut = 15000)
```

Comment:
Each invocation takes 5s+, 10 invocations would add 1min runtime to the test suite. Can we do a single invoke and reduce warmup duration to something like 2s?

Reply:
Resolved.